includes/class-order.php

```diff
@@ -219,24 +324,186 @@ public static function update_status( $order_id, $response ) {
+	 * @return bool
+	 */
+	public static function handle_callback( $order_id, $callback ) {
+
+		// Check callback.
+		if ( empty( $callback ) ) {
 			$error = sprintf(
 				// translators: %s WC Order ID.
-				esc_html__( 'Order status response empty for WC Order #%s.', 'zota-woocommerce' ),
+				esc_html__( 'Order callback empty for WC Order #%s.', 'zota-woocommerce' ),
 				(int) $order_id
 			);
 			Zotapay::getLogger()->error( $error );
 			return false;
 		}
 
-		// If no change do nothing.
-		if ( $order->get_meta( '_zotapay_status', true ) === $response->getStatus() ) {
+		$response['status']                 = $callback->getStatus();
+		$response['processorTransactionID'] = $callback->getProcessorTransactionID();
+		$response['errorMessage']           = $callback->getErrorMessage();
+
+		if ( self::amount_changed( $callback ) ) {
+			$response['amountChanged']  = true;
+			$response['amount']         = $callback->getAmount();
+			$response['originalAmount'] = $extra_data['originalAmount'];
+		}
+
+		return self::update_status( $order_id, $response );
+	}
+
+
+	/**
+	 * Handle amount changed
+	 *
+	 * @param  WC_order $order           WC Order.
+	 * @param  string   $amount          Amount paid.
+	 * @param  string   $original_amount Original order amount.
+	 * @return bool
+	 */
+	public static function handle_amount_changed( $order, $amount, $original_amount ) {
+
+		// Add meta.
+		$order->add_meta_data( '_zotapay_amount_changed', sanitize_text_field( $amount ) );
+
+		// Order note.
+		if ( \floatval( $amount ) < \floatval( $original_amount ) ) {
+			$note = sprintf(
+				// translators: %1$s amount paid, %2$s original order amount.
+				esc_html__( 'ZotaPay order partial payment. %1$s of %2$s paid.', 'zota-woocommerce' ),
+				sanitize_text_field( $amount ),
+				sanitize_text_field( $original_amount )
+			);
+		} else {
+			$note = sprintf(
+				// translators: %1$s amount paid, %2$s original order amount.
+				esc_html__( 'ZotaPay order overpayment. %1$s of %2$s paid.', 'zota-woocommerce' ),
+				sanitize_text_field( $amount ),
+				sanitize_text_field( $original_amount )
+			);
+		}
+		$order->add_order_note( $note );
+		$order->save();
+
+		// Get total paid.
+		$total_paid = self::get_total_paid( $order );
+
+		// If total paid is lower than original amount set order status to Partial Payment.
+		if ( \floatval( $total_paid ) < \floatval( $original_amount ) ) {
+			$order->set_status( 'wc-partial-payment' );
+			$order->save();
+			return true;
+		}
+
+		// If total paid is bigger than original amount set order status to Overpaid.
+		if ( \floatval( $total_paid ) > \floatval( $original_amount ) ) {
+			$order->set_date_paid( time() );
+			$order->set_status( 'wc-overpayment' );
+			$order->save();
+			return true;
+		}
+
+		// If total paid is equal to original amount set order payment complete.
+		if ( \floatval( $total_paid ) === \floatval( $original_amount ) ) {
+			$order->payment_complete();
+			return true;
+		}
```

Comment:
Can we avoid the repetitve `floatval` calls here? And maybe use `if () elseif() else` conditional structure?

Reply:
Yap.